Compatibility with Python 3.15

Hi! It's me again. We've started testing pure_eval with Python 3.15 already in Fedora and 11 tests failed:

The log from pytest is:
<details>

> ============================= test session starts ==============================
> platform linux -- Python 3.15.0a3, pytest-8.4.2, pluggy-1.6.0
> cachedir: .tox/py315/.pytest_cache
> rootdir: /builddir/build/BUILD/python-pure-eval-0.2.3-build/pure_eval-0.2.3
> configfile: pyproject.toml
> collected 47 items
> 
> tests/test_core.py ..F.....F........                                     [ 36%]
> tests/test_getattr_static.py F..FFF..F.FF.F...F.....                     [ 85%]
> tests/test_utils.py .......                                              [100%]
> 
> =================================== FAILURES ===================================
> _______________________________ test_eval_attrs ________________________________
> 
>     def test_eval_attrs():
>         class Foo:
>             bar = 9
>     
>             @property
>             def prop(self):
>                 return 0
>     
>             def method(self):
>                 pass
>     
>         foo = Foo()
>         foo.spam = 44
>     
> >       check_eval(
>             "foo.bar + foo.spam + Foo.bar",
>             foo.bar, foo.spam, Foo.bar,
>             foo.bar + foo.spam,
>             foo.bar + foo.spam + Foo.bar,
>             foo, Foo
>         )
> 
> tests/test_core.py:63: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> 
> source = 'foo.bar + foo.spam + Foo.bar', total = True
> expected_values = (9, 44, 9, 53, 62, <tests.test_core.test_eval_attrs.<locals>.Foo object at 0x7f83821d6ba0>, ...)
> frame = <frame at 0x7f838243b120, file '/builddir/build/BUILD/python-pure-eval-0.2.3-build/pure_eval-0.2.3/tests/test_core.py', line 63, code test_eval_attrs>
> evaluator = <pure_eval.core.Evaluator object at 0x7f8381994f50>
> 
>     def check_eval(source, *expected_values, total=True):
>         frame = inspect.currentframe().f_back
>         evaluator = Evaluator.from_frame(frame)
>         root = ast.parse(source)
>         values = []
>         for node, value in evaluator.find_expressions(root):
>             expr = ast.Expression(body=node)
>             ast.copy_location(expr, node)
>             code = compile(expr, "<expr>", "eval")
>             expected = eval(code, frame.f_globals, frame.f_locals)
>             assert value == expected
>             values.append(value)
>             if total:
>                 assert value in expected_values
>     
>         for expected in expected_values:
> >           assert expected in values
> E           AssertionError: assert 44 in [9, <class 'tests.test_core.test_eval_attrs.<locals>.Foo'>, <tests.test_core.test_eval_attrs.<locals>.Foo object at 0x7f83821d6ba0>, <tests.test_core.test_eval_attrs.<locals>.Foo object at 0x7f83821d6ba0>]
> 
> tests/test_core.py:29: AssertionError
> ________________________ test_is_expression_interesting ________________________
> 
>     def test_is_expression_interesting():
>         class Foo:
>             def method(self):
>                 pass
>     
>             alias = method
>     
>         foo = Foo()
>         x = [check_interesting]
>         foo.x = x
>         assert check_interesting('x')
>         assert not check_interesting('help')
>         assert not check_interesting('check_interesting')
>         assert not check_interesting('[1]')
>         assert check_interesting('[1, 2][0]')
>         assert check_interesting('foo')
>         assert not check_interesting('Foo')
> >       assert check_interesting('foo.x')
>                ^^^^^^^^^^^^^^^^^^^^^^^^^^
> 
> tests/test_core.py:293: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> tests/test_core.py:269: in check_interesting
>     raise cannot
> tests/test_core.py:253: in check_interesting
>     value = evaluator[node]
>             ^^^^^^^^^^^^^^^
> pure_eval/core.py:69: in __getitem__
>     self._cache[node] = result = self._handle(node)
>                                  ^^^^^^^^^^^^^^^^^^
> pure_eval/core.py:96: in _handle
>     return getattr_static(value, attr)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> 
> obj = <tests.test_core.test_is_expression_interesting.<locals>.Foo object at 0x7f83821d7b60>
> attr = 'x'
> 
>     def getattr_static(obj, attr):
>         """Retrieve attributes without triggering dynamic lookup via the
>            descriptor protocol,  __getattr__ or __getattribute__.
>     
>            Note: this function may not be able to retrieve all attributes
>            that getattr can fetch (like dynamically created attributes)
>            and may find attributes that getattr can't (like descriptors
>            that raise AttributeError). It can also return descriptor objects
>            instead of instance members in some cases. See the
>            documentation for details.
>         """
>         instance_result = _sentinel
>         if not _is_type(obj):
>             klass = type(obj)
>             dict_attr = _shadowed_dict(klass)
>             if (dict_attr is _sentinel or
>                     type(dict_attr) is types.MemberDescriptorType):
>                 instance_result = _check_instance(obj, attr)
>             else:
> >               raise CannotEval
> E               pure_eval.utils.CannotEval: CannotEval
> 
> pure_eval/my_getattr_static.py:75: CannotEval
> _________________________ TestGetattrStatic.test_basic _________________________
> 
> self = <tests.test_getattr_static.TestGetattrStatic testMethod=test_basic>
> 
>     def test_basic(self):
>         class Thing(object):
>             x = object()
>     
>         thing = Thing()
> >       self.assert_getattr(thing, 'x')
> 
> tests/test_getattr_static.py:27: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> tests/test_getattr_static.py:14: in assert_getattr
>     getattr_static(thing, attr),
>     ^^^^^^^^^^^^^^^^^^^^^^^^^^^
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> 
> obj = <tests.test_getattr_static.TestGetattrStatic.test_basic.<locals>.Thing object at 0x7f838168f0e0>
> attr = 'x'
> 
>     def getattr_static(obj, attr):
>         """Retrieve attributes without triggering dynamic lookup via the
>            descriptor protocol,  __getattr__ or __getattribute__.
>     
>            Note: this function may not be able to retrieve all attributes
>            that getattr can fetch (like dynamically created attributes)
>            and may find attributes that getattr can't (like descriptors
>            that raise AttributeError). It can also return descriptor objects
>            instead of instance members in some cases. See the
>            documentation for details.
>         """
>         instance_result = _sentinel
>         if not _is_type(obj):
>             klass = type(obj)
>             dict_attr = _shadowed_dict(klass)
>             if (dict_attr is _sentinel or
>                     type(dict_attr) is types.MemberDescriptorType):
>                 instance_result = _check_instance(obj, attr)
>             else:
> >               raise CannotEval
> E               pure_eval.utils.CannotEval: CannotEval
> 
> pure_eval/my_getattr_static.py:75: CannotEval
> ___________________ TestGetattrStatic.test_class_as_property ___________________
> 
> self = <tests.test_getattr_static.TestGetattrStatic testMethod=test_class_as_property>
> 
>     def test_class_as_property(self):
>         class Base(object):
>             foo = 3
>     
>         class Something(Base):
>             @property
>             def __class__(self):
>                 return 1 / 0
>     
>         instance = Something()
> >       self.assert_getattr(instance, 'foo')
> 
> tests/test_getattr_static.py:217: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> tests/test_getattr_static.py:14: in assert_getattr
>     getattr_static(thing, attr),
>     ^^^^^^^^^^^^^^^^^^^^^^^^^^^
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> 
> obj = <tests.test_getattr_static.TestGetattrStatic.test_class_as_property.<locals>.Something object at 0x7f838168ee40>
> attr = 'foo'
> 
>     def getattr_static(obj, attr):
>         """Retrieve attributes without triggering dynamic lookup via the
>            descriptor protocol,  __getattr__ or __getattribute__.
>     
>            Note: this function may not be able to retrieve all attributes
>            that getattr can fetch (like dynamically created attributes)
>            and may find attributes that getattr can't (like descriptors
>            that raise AttributeError). It can also return descriptor objects
>            instead of instance members in some cases. See the
>            documentation for details.
>         """
>         instance_result = _sentinel
>         if not _is_type(obj):
>             klass = type(obj)
>             dict_attr = _shadowed_dict(klass)
>             if (dict_attr is _sentinel or
>                     type(dict_attr) is types.MemberDescriptorType):
>                 instance_result = _check_instance(obj, attr)
>             else:
> >               raise CannotEval
> E               pure_eval.utils.CannotEval: CannotEval
> 
> pure_eval/my_getattr_static.py:75: CannotEval
> __________________ TestGetattrStatic.test_custom_object_dict ___________________
> 
> self = <tests.test_getattr_static.TestGetattrStatic testMethod=test_custom_object_dict>
> 
>     def test_custom_object_dict(self):
>         class Custom(dict):
>             def get(self, key, default=None):
>                 return 1 / 0
>     
>             __getitem__ = get
>     
>         class Foo(object):
>             a = 3
>     
>         foo = Foo()
>         foo.__dict__ = Custom()
>         foo.x = 5
> >       self.assert_getattr(foo, 'a')
> 
> tests/test_getattr_static.py:261: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> tests/test_getattr_static.py:14: in assert_getattr
>     getattr_static(thing, attr),
>     ^^^^^^^^^^^^^^^^^^^^^^^^^^^
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> 
> obj = <tests.test_getattr_static.TestGetattrStatic.test_custom_object_dict.<locals>.Foo object at 0x7f838168ecf0>
> attr = 'a'
> 
>     def getattr_static(obj, attr):
>         """Retrieve attributes without triggering dynamic lookup via the
>            descriptor protocol,  __getattr__ or __getattribute__.
>     
>            Note: this function may not be able to retrieve all attributes
>            that getattr can fetch (like dynamically created attributes)
>            and may find attributes that getattr can't (like descriptors
>            that raise AttributeError). It can also return descriptor objects
>            instead of instance members in some cases. See the
>            documentation for details.
>         """
>         instance_result = _sentinel
>         if not _is_type(obj):
>             klass = type(obj)
>             dict_attr = _shadowed_dict(klass)
>             if (dict_attr is _sentinel or
>                     type(dict_attr) is types.MemberDescriptorType):
>                 instance_result = _check_instance(obj, attr)
>             else:
> >               raise CannotEval
> E               pure_eval.utils.CannotEval: CannotEval
> 
> pure_eval/my_getattr_static.py:75: CannotEval
> ______________________ TestGetattrStatic.test_descriptor _______________________
> 
> self = <tests.test_getattr_static.TestGetattrStatic testMethod=test_descriptor>
> 
>     def test_descriptor(self):
>         class descriptor(object):
>             def __get__(self, instance, owner):
>                 return 3
>     
>         class Foo(object):
>             d = descriptor()
>     
>         foo = Foo()
>     
>         # for a non data descriptor we return the instance attribute
>         foo.__dict__['d'] = 1
> >       self.assert_getattr(foo, 'd')
> 
> tests/test_getattr_static.py:184: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> tests/test_getattr_static.py:14: in assert_getattr
>     getattr_static(thing, attr),
>     ^^^^^^^^^^^^^^^^^^^^^^^^^^^
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> 
> obj = <tests.test_getattr_static.TestGetattrStatic.test_descriptor.<locals>.Foo object at 0x7f838168ea50>
> attr = 'd'
> 
>     def getattr_static(obj, attr):
>         """Retrieve attributes without triggering dynamic lookup via the
>            descriptor protocol,  __getattr__ or __getattribute__.
>     
>            Note: this function may not be able to retrieve all attributes
>            that getattr can fetch (like dynamically created attributes)
>            and may find attributes that getattr can't (like descriptors
>            that raise AttributeError). It can also return descriptor objects
>            instead of instance members in some cases. See the
>            documentation for details.
>         """
>         instance_result = _sentinel
>         if not _is_type(obj):
>             klass = type(obj)
>             dict_attr = _shadowed_dict(klass)
>             if (dict_attr is _sentinel or
>                     type(dict_attr) is types.MemberDescriptorType):
>                 instance_result = _check_instance(obj, attr)
>             else:
> >               raise CannotEval
> E               pure_eval.utils.CannotEval: CannotEval
> 
> pure_eval/my_getattr_static.py:75: CannotEval
> _______________________ TestGetattrStatic.test_inherited _______________________
> 
> self = <tests.test_getattr_static.TestGetattrStatic testMethod=test_inherited>
> 
>     def test_inherited(self):
>         class Thing(object):
>             x = object()
>     
>         class OtherThing(Thing):
>             pass
>     
>         something = OtherThing()
> >       self.assert_getattr(something, 'x')
> 
> tests/test_getattr_static.py:38: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> tests/test_getattr_static.py:14: in assert_getattr
>     getattr_static(thing, attr),
>     ^^^^^^^^^^^^^^^^^^^^^^^^^^^
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> 
> obj = <tests.test_getattr_static.TestGetattrStatic.test_inherited.<locals>.OtherThing object at 0x7f838168c1a0>
> attr = 'x'
> 
>     def getattr_static(obj, attr):
>         """Retrieve attributes without triggering dynamic lookup via the
>            descriptor protocol,  __getattr__ or __getattribute__.
>     
>            Note: this function may not be able to retrieve all attributes
>            that getattr can fetch (like dynamically created attributes)
>            and may find attributes that getattr can't (like descriptors
>            that raise AttributeError). It can also return descriptor objects
>            instead of instance members in some cases. See the
>            documentation for details.
>         """
>         instance_result = _sentinel
>         if not _is_type(obj):
>             klass = type(obj)
>             dict_attr = _shadowed_dict(klass)
>             if (dict_attr is _sentinel or
>                     type(dict_attr) is types.MemberDescriptorType):
>                 instance_result = _check_instance(obj, attr)
>             else:
> >               raise CannotEval
> E               pure_eval.utils.CannotEval: CannotEval
> 
> pure_eval/my_getattr_static.py:75: CannotEval
> ____________________ TestGetattrStatic.test_inherited_slots ____________________
> 
> self = <tests.test_getattr_static.TestGetattrStatic testMethod=test_inherited_slots>
> 
>     def test_inherited_slots(self):
>         class Thing(object):
>             __slots__ = ['x']
>     
>             def __init__(self):
>                 self.x = 'foo'
>     
>         class OtherThing(Thing):
>             pass
>     
> >       self.assert_getattr(OtherThing(), 'x')
> 
> tests/test_getattr_static.py:170: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> tests/test_getattr_static.py:14: in assert_getattr
>     getattr_static(thing, attr),
>     ^^^^^^^^^^^^^^^^^^^^^^^^^^^
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> 
> obj = <tests.test_getattr_static.TestGetattrStatic.test_inherited_slots.<locals>.OtherThing object at 0x7f838168c2f0>
> attr = 'x'
> 
>     def getattr_static(obj, attr):
>         """Retrieve attributes without triggering dynamic lookup via the
>            descriptor protocol,  __getattr__ or __getattribute__.
>     
>            Note: this function may not be able to retrieve all attributes
>            that getattr can fetch (like dynamically created attributes)
>            and may find attributes that getattr can't (like descriptors
>            that raise AttributeError). It can also return descriptor objects
>            instead of instance members in some cases. See the
>            documentation for details.
>         """
>         instance_result = _sentinel
>         if not _is_type(obj):
>             klass = type(obj)
>             dict_attr = _shadowed_dict(klass)
>             if (dict_attr is _sentinel or
>                     type(dict_attr) is types.MemberDescriptorType):
>                 instance_result = _check_instance(obj, attr)
>             else:
> >               raise CannotEval
> E               pure_eval.utils.CannotEval: CannotEval
> 
> pure_eval/my_getattr_static.py:75: CannotEval
> _____________________ TestGetattrStatic.test_instance_attr _____________________
> 
> self = <tests.test_getattr_static.TestGetattrStatic testMethod=test_instance_attr>
> 
>     def test_instance_attr(self):
>         class Thing(object):
>             x = 2
>     
>             def __init__(self, x):
>                 self.x = x
>     
>         thing = Thing(3)
> >       self.assert_getattr(thing, 'x')
> 
> tests/test_getattr_static.py:48: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> tests/test_getattr_static.py:14: in assert_getattr
>     getattr_static(thing, attr),
>     ^^^^^^^^^^^^^^^^^^^^^^^^^^^
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> 
> obj = <tests.test_getattr_static.TestGetattrStatic.test_instance_attr.<locals>.Thing object at 0x7f838168c440>
> attr = 'x'
> 
>     def getattr_static(obj, attr):
>         """Retrieve attributes without triggering dynamic lookup via the
>            descriptor protocol,  __getattr__ or __getattribute__.
>     
>            Note: this function may not be able to retrieve all attributes
>            that getattr can fetch (like dynamically created attributes)
>            and may find attributes that getattr can't (like descriptors
>            that raise AttributeError). It can also return descriptor objects
>            instead of instance members in some cases. See the
>            documentation for details.
>         """
>         instance_result = _sentinel
>         if not _is_type(obj):
>             klass = type(obj)
>             dict_attr = _shadowed_dict(klass)
>             if (dict_attr is _sentinel or
>                     type(dict_attr) is types.MemberDescriptorType):
>                 instance_result = _check_instance(obj, attr)
>             else:
> >               raise CannotEval
> E               pure_eval.utils.CannotEval: CannotEval
> 
> pure_eval/my_getattr_static.py:75: CannotEval
> ______________ TestGetattrStatic.test_metaclass_dict_as_property _______________
> 
> self = <tests.test_getattr_static.TestGetattrStatic testMethod=test_metaclass_dict_as_property>
> 
>     def test_metaclass_dict_as_property(self):
>         class Meta(type):
>             @property
>             def __dict__(self):
>                 return 1 / 0
>     
>         class Thing(metaclass=Meta):
>             bar = 4
>     
>             def __init__(self):
>                 self.spam = 42
>     
>         instance = Thing()
> >       self.assert_getattr(instance, "spam")
> 
> tests/test_getattr_static.py:277: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> tests/test_getattr_static.py:14: in assert_getattr
>     getattr_static(thing, attr),
>     ^^^^^^^^^^^^^^^^^^^^^^^^^^^
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> 
> obj = <tests.test_getattr_static.TestGetattrStatic.test_metaclass_dict_as_property.<locals>.Thing object at 0x7f838168c590>
> attr = 'spam'
> 
>     def getattr_static(obj, attr):
>         """Retrieve attributes without triggering dynamic lookup via the
>            descriptor protocol,  __getattr__ or __getattribute__.
>     
>            Note: this function may not be able to retrieve all attributes
>            that getattr can fetch (like dynamically created attributes)
>            and may find attributes that getattr can't (like descriptors
>            that raise AttributeError). It can also return descriptor objects
>            instead of instance members in some cases. See the
>            documentation for details.
>         """
>         instance_result = _sentinel
>         if not _is_type(obj):
>             klass = type(obj)
>             dict_attr = _shadowed_dict(klass)
>             if (dict_attr is _sentinel or
>                     type(dict_attr) is types.MemberDescriptorType):
>                 instance_result = _check_instance(obj, attr)
>             else:
> >               raise CannotEval
> E               pure_eval.utils.CannotEval: CannotEval
> 
> pure_eval/my_getattr_static.py:75: CannotEval
> ____________________ TestGetattrStatic.test_mro_as_property ____________________
> 
> self = <tests.test_getattr_static.TestGetattrStatic testMethod=test_mro_as_property>
> 
>     def test_mro_as_property(self):
>         class Meta(type):
>             @property
>             def __mro__(self):
>                 return 1 / 0
>     
>         class Base(object):
>             foo = 3
>     
>         class Something(Base, metaclass=Meta):
>             pass
>     
> >       self.assert_getattr(Something(), 'foo')
> 
> tests/test_getattr_static.py:232: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> tests/test_getattr_static.py:14: in assert_getattr
>     getattr_static(thing, attr),
>     ^^^^^^^^^^^^^^^^^^^^^^^^^^^
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> 
> obj = <tests.test_getattr_static.TestGetattrStatic.test_mro_as_property.<locals>.Something object at 0x7f838168c830>
> attr = 'foo'
> 
>     def getattr_static(obj, attr):
>         """Retrieve attributes without triggering dynamic lookup via the
>            descriptor protocol,  __getattr__ or __getattribute__.
>     
>            Note: this function may not be able to retrieve all attributes
>            that getattr can fetch (like dynamically created attributes)
>            and may find attributes that getattr can't (like descriptors
>            that raise AttributeError). It can also return descriptor objects
>            instead of instance members in some cases. See the
>            documentation for details.
>         """
>         instance_result = _sentinel
>         if not _is_type(obj):
>             klass = type(obj)
>             dict_attr = _shadowed_dict(klass)
>             if (dict_attr is _sentinel or
>                     type(dict_attr) is types.MemberDescriptorType):
>                 instance_result = _check_instance(obj, attr)
>             else:
> >               raise CannotEval
> E               pure_eval.utils.CannotEval: CannotEval
> 
> pure_eval/my_getattr_static.py:75: CannotEval
> =========================== short test summary info ============================
> FAILED tests/test_core.py::test_eval_attrs - AssertionError: assert 44 in [9,...
> FAILED tests/test_core.py::test_is_expression_interesting - pure_eval.utils.C...
> FAILED tests/test_getattr_static.py::TestGetattrStatic::test_basic - pure_eva...
> FAILED tests/test_getattr_static.py::TestGetattrStatic::test_class_as_property
> FAILED tests/test_getattr_static.py::TestGetattrStatic::test_custom_object_dict
> FAILED tests/test_getattr_static.py::TestGetattrStatic::test_descriptor - pur...
> FAILED tests/test_getattr_static.py::TestGetattrStatic::test_inherited - pure...
> FAILED tests/test_getattr_static.py::TestGetattrStatic::test_inherited_slots
> FAILED tests/test_getattr_static.py::TestGetattrStatic::test_instance_attr - ...
> FAILED tests/test_getattr_static.py::TestGetattrStatic::test_metaclass_dict_as_property
> FAILED tests/test_getattr_static.py::TestGetattrStatic::test_mro_as_property
> ======================== 11 failed, 36 passed in 1.53s =========================
> py315: exit 1 (1.73 seconds) /builddir/build/BUILD/python-pure-eval-0.2.3-build/pure_eval-0.2.3> pytest pid=266
>   py315: FAIL code 1 (1.73 seconds)
>   evaluation failed :( (1.76 seconds)
> error: Bad exit status from /var/tmp/rpm-tmp.TG1l3H (%check)
>     Bad exit status from /var/tmp/rpm-tmp.TG1l3H (%check)

</details>

--- a/.github/workflows/pytest.yml
+++ b/.github/workflows/pytest.yml
@@ -12,7 +12,7 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        python-version: [3.8, 3.9, '3.10', 3.11, 3.12, 3.13, 3.14]
+        python-version: [3.8, 3.9, '3.10', 3.11, 3.12, 3.13, 3.14, 3.15-dev]
     steps:
     - uses: actions/checkout@v4
     - name: Set up Python ${{ matrix.python-version }}

--- a/pure_eval/my_getattr_static.py
+++ b/pure_eval/my_getattr_static.py
@@ -69,7 +69,7 @@ def getattr_static(obj, attr):
         klass = type(obj)
         dict_attr = _shadowed_dict(klass)
         if (dict_attr is _sentinel or
-                type(dict_attr) is types.MemberDescriptorType):
+                type(dict_attr) is types.MemberDescriptorType or type(dict_attr) is types.GetSetDescriptorType):
             instance_result = _check_instance(obj, attr)
         else:
             raise CannotEval

--- a/setup.cfg
+++ b/setup.cfg
@@ -16,6 +16,7 @@ classifiers =
     Programming Language :: Python :: 3.12
     Programming Language :: Python :: 3.13
     Programming Language :: Python :: 3.14
+    Programming Language :: Python :: 3.15
     License :: OSI Approved :: MIT License
     Operating System :: OS Independent
 

--- a/tox.ini
+++ b/tox.ini
@@ -1,5 +1,5 @@
 [tox]
-envlist = py{38,39,310,311,312,313,314}
+envlist = py{38,39,310,311,312,313,314,315}
 
 [testenv]
 commands = pytest

